File: AllSamples/Samples/Samples_Source/natural language/speller/vb/cmdopts.vb
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------
Imports System
Imports System.Collections
Imports System.Globalization


Namespace Microsoft.Samples.NaturalLanguage
   Public Class CmdOpts
      
      Public Sub New()
         m_sInput = ""
         m_sOutput = ""
         m_Culture = Nothing
         m_fVerbose = False
         m_LexList = New ArrayList()
      End Sub 'New
      
      Public Property Output() As String
         Get
            Return m_sOutput
         End Get
         Set
            m_sOutput = value
         End Set
      End Property 
      
      Public Property Input() As String
         Get
            Return m_sInput
         End Get
         Set
            m_sInput = value
         End Set
      End Property 
      
      Public Property Culture() As CultureInfo
         Get
            Return m_Culture
         End Get
         Set
            m_Culture = value
         End Set
      End Property 
      
      Public Property Verbose() As Boolean
         Get
            Return m_fVerbose
         End Get
         Set
            m_fVerbose = value
         End Set
      End Property 
      
      Public ReadOnly Property LexiconList() As ArrayList
         Get
            Return m_LexList
         End Get
      End Property
       
      Public Sub DisplayHelp()
         DisplayHelpLine("I", "Input", "The text file or directory to use for intput text (defaults to standard input)")
         DisplayHelpLine("O", "Output", "The output file (Defaults to Standard Out)")
         DisplayHelpLine("C", "Culture", "Sets the culture of the engine (ex. -C 'en-us' or 'fr-ca').  Defaults to current culture. ")
         DisplayHelpLine("V", "Verbose", "Verbose output")
         DisplayHelpLine("LEX", "Lexicon", "Adds a lexicon to the speller engine")
      End Sub 'DisplayHelp
       ' DisplayHelp
      Public Function ProcessOpts(args() As String) As Boolean
         Dim invariantCulture As CultureInfo = CultureInfo.InvariantCulture
         
         Dim i As Integer
         
         While i < args.Length
            If args(i).StartsWith("-") Then
               Dim thisarg As String = args(i)
               
               thisarg = thisarg.Remove(0, 1).ToLower(System.Globalization.CultureInfo.CurrentCulture)
               i = i + 1
               If thisarg = "input" Or thisarg = "i" Then
                  If i >= args.Length Then
                     Throw New System.ArgumentException([String].Concat("No Value for ", thisarg))
                  End If
                  
                  ' assign to the next arg
                  m_sInput = args(i)
               End If
               
               If thisarg = "output" Or thisarg = "o" Then
                  i = i + 1
                  If i >= args.Length Then
                     Throw New System.ArgumentException([String].Concat("No Value for ", thisarg))
                  End If
                  
                  ' assign to the next arg
                  m_sOutput = args(i)
               End If
               
               If thisarg = "culture" Or thisarg = "c" Then
                  i = i + 1
                  If i >= args.Length Then
                     Throw New System.ArgumentException("No Value for ", thisarg)
                  End If
                  
                  ' assign to the next arg
                  m_Culture = GetCultureFromArguments(args(i))
               End If
               
               If [String].Compare(thisarg, "lexicon", True, invariantCulture) = 0 Or [String].Compare(thisarg, "lex", True, invariantCulture) = 0 Then
                  i = i + 1
                  If i >= args.Length Then
                     Throw New System.ArgumentException("No value for ", thisarg)
                  End If
                  
                  m_LexList.Add(New System.NaturalLanguage.Lexicon(args(i)))
               End If
               
               If thisarg = "verbose" Or thisarg = "v" Then
                  m_fVerbose = True
               End If
               
               If thisarg = "help" Or thisarg = "?" Then
                  DisplayHelp()
                  Return False
               End If
            End If
         End While ' for
         Return True
      End Function 'ProcessOpts
       ' ProcessOpts
      Private Sub DisplayHelpLine(shortname As String, name As String, description As String)
         Console.WriteLine(" -{0,1} {1,-24} {2}", shortname, name, description)
      End Sub 'DisplayHelpLine
      
      
      Private Function GetCultureFromArguments(arguments As String) As CultureInfo
         Dim culture As CultureInfo = CultureInfo.CurrentCulture
         
         Try
            culture = New CultureInfo(arguments)
            Return culture
         Catch e As ArgumentException
            Console.Error.WriteLine(e.Message)
            Try
               culture = New CultureInfo(Int32.Parse(arguments, CultureInfo.CurrentCulture.NumberFormat), True)
               Return culture
            Catch
            Catch ex As ArgumentException
               Console.Error.WriteLine(ex.Message)
               Return CultureInfo.CurrentCulture
            End Try
         End Try
      End Function 'GetCultureFromArguments
      
      Private m_sInput As String
      
      Private m_sOutput As String
      
      Private m_Culture As CultureInfo
      
      Private m_LexList As ArrayList
      
      Private m_fVerbose As Boolean
   End Class ' class CmdOpts
End Namespace 'Microsoft.Samples.NaturalLanguage

File: AllSamples/Samples/Samples_Source/natural language/speller/csharp/cmdopts.cs
//---------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.Collections;
using System.Globalization;

namespace Microsoft.Samples.NaturalLanguage
{
	public class CmdOpts
	{
		public CmdOpts()
		{
			m_sInput = "";
			m_sOutput = "";
			m_Culture = null;
			m_fVerbose = false;
			m_LexList = new ArrayList();
		}

		public string Output
		{
			get { return m_sOutput; }
			set { m_sOutput = value; }
		}

		public string Input
		{
			get { return m_sInput; }
			set { m_sInput = value; }
		}

		public CultureInfo Culture
		{
			get { return m_Culture; }
			set { m_Culture = value; }
		}

		public bool Verbose
		{
			get { return m_fVerbose; }
			set { m_fVerbose = value; }
		}

		public ArrayList LexiconList
		{
			get { return m_LexList; }
		}

		public void DisplayHelp()
		{
			DisplayHelpLine("I", "Input", "The text file or directory to use for intput text (defaults to standard input)");
			DisplayHelpLine("O", "Output", "The output file (Defaults to Standard Out)");
			DisplayHelpLine("C", "Culture", "Sets the culture of the engine (ex. -C 'en-us' or 'fr-ca').  Defaults to current culture. ");
			DisplayHelpLine("V", "Verbose", "Verbose output");
			DisplayHelpLine("LEX", "Lexicon", "Adds a lexicon to the speller engine");
		} // DisplayHelp

		public bool ProcessOpts(string[] args)
		{
			CultureInfo invariantCulture = CultureInfo.InvariantCulture;

			for (int i = 0; i < args.Length; ++i)
			{
				if (args[i].StartsWith("-"))
				{
					string thisarg = args[i];

					thisarg = thisarg.Remove(0, 1).ToLower(System.Globalization.CultureInfo.CurrentCulture);
					if (thisarg == "input" || thisarg =="i" )
					{
						if (++i >= args.Length)
						{   // make sure there's a next arg to move on to
							throw new System.ArgumentException(String.Concat("No Value for ", thisarg));
						}

						// assign to the next arg
						m_sInput = args[i];
					}

					if (thisarg == "output" || thisarg == "o")
					{
						if (++i >= args.Length)
						{   // make sure there's a next arg to move on to
							throw new System.ArgumentException(String.Concat("No Value for ", thisarg));
						}

						// assign to the next arg
						m_sOutput = args[i];
					}

					if (thisarg == "culture" || thisarg == "c")
					{
						if (++i >= args.Length)
						{   // make sure there's a next arg to move on to
							throw new System.ArgumentException("No Value for ", thisarg);
						}

						// assign to the next arg
						m_Culture = GetCultureFromArguments(args[i]);
					}

					if ((String.Compare(thisarg, "lexicon", true, invariantCulture) == 0) || (String.Compare(thisarg, "lex", true, invariantCulture) == 0))
					{
						if (++i >= args.Length)
						{
							throw new System.ArgumentException("No value for ", thisarg);
						}

						m_LexList.Add(new System.NaturalLanguage.Lexicon(args[i]));
					}

					if (thisarg == "verbose" || thisarg == "v")
					{
						m_fVerbose = true;
					}

					if (thisarg == "help" || thisarg == "?")
					{
						DisplayHelp();
						return false;
					}
				}
			} // for

			return true;
		} // ProcessOpts

		private void DisplayHelpLine(string shortname, string name, string description)
		{
			Console.WriteLine(" -{0,1} {1,-24} {2}", shortname, name, description);
		}

		private CultureInfo GetCultureFromArguments(string arguments)
		{
			CultureInfo culture = CultureInfo.CurrentCulture;

			try
			{
				culture = new CultureInfo(arguments);
				return culture;
			}
			catch (ArgumentException e)
			{
				Console.Error.WriteLine(e.Message);
				try
				{
					culture = new CultureInfo(Int32.Parse(arguments, CultureInfo.CurrentCulture.NumberFormat), true);
					return culture;
				}
				catch (FormatException)
				{
					Console.Error.WriteLine("defaulting to current culture.");
					return CultureInfo.CurrentCulture;
				}
				catch (ArgumentException ex)
				{
					Console.Error.WriteLine(ex.Message);
					return CultureInfo.CurrentCulture;
				}
			}
		}

		private string m_sInput;

		private string m_sOutput;

		private CultureInfo m_Culture;

		private ArrayList m_LexList;

		private bool m_fVerbose;
	}; // class CmdOpts
}
